### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by fepitre**
Need to be checked and reviewed.

### CODE REVIEW

File: build-infra/build-vm.sls [hunk 1]
```diff
@@ -144,3 +165,96 @@ commands-keyring:
...
+/home/user/.ssh/config:
+  file.managed:
+    - contents: |
+{% for host, config in salt['pillar.get']('build-infra:remote-hosts', {}).items() %}
+        Host {{host}}
+          HostName {{host}}
+          User {{config.ssh_user}}
+{% endfor %}
+    - mode: 0755
+    - makedirs: True
```
Comment (marmarek) [inline build-infra/build-vm.sls:179]: Missing `user: user`

File: build-infra/build-vm.sls [hunk 2]
```diff
@@ -144,3 +165,96 @@ commands-keyring:
...
+{{builder}}-init:
+  cmd.run:
+    - name: "BUILDERCONF= GIT_URL_builder=https://github.com/QubesOS/qubes-builder COMPONENTS=builder make get-sources"
+    - cwd: {{ builder }}
+    - runas: user
+    - require:
+      - cmd: {{builder}}-check
+
+{% for key in keys %}
+{{builder}}-{{key}}:
```
Comment (marmarek) [inline build-infra/build-vm.sls]: Since `setup` script will need those keys too, I'd put them into qubes-builder repository. This also means you'll have there already.

File: build-infra/build-vm.sls [hunk 3]
```diff
@@ -144,3 +165,96 @@ commands-keyring:
...
+  file.managed:
+    - name: /home/user/{{key}}
+    - source: salt://build-infra/keys/{{key}}
+    - user: user
+
+{{builder}}-{{key}}-import:
+  cmd.run:
+    - name:  export GNUPGHOME="$(make get-var GET_VAR=KEYRING_DIR_GIT)"; scripts/verify-git-tag; gpg --import /home/user/{{key}} || exit 1; echo '{{key}}:6:' | gpg --import-ownertrust
```
Comment (marmarek) [inline build-infra/build-vm.sls]: `scripts/verify-git-tag` call is only to create initial keyring. Since you call `get-sources` above, it should be there already.

File: pillar/build-infra/init.sls [hunk 4]
```diff
@@ -1,21 +1,61 @@
...
+                - baseurl: https://github.com/QubesOS/qubes-
+                - component: release-configs
+      volume-size: 20GiB
+
+# Above definitions of environments can be put in
+# separate pillar files. Just ensure to keep at least
+# the env names for init.top
+#   build-envs:
+#     fedora1
+#     fedora2
```
Comment (marmarek) [inline pillar/build-infra/init.sls]: Syntax looks wrong, probably you wanted `fedora1: {}`.

File: build-infra/build-logs.sls [hunk 5]
```diff
@@ -1,4 +1,7 @@
-{% set build_log_key_fpr = '1B760CED53D8EB5AE529BEBB16D7539A228D30DB' %}
+{% set build_log_key_fpr = salt('pillar.get')('build-infra:build_log_key_fpr', '1B760CED53D8EB5AE529BEBB16D7539A228D30DB') %}
+{% set build_log_repo = salt('pillar.get')('build-infra:build_log_repo', 'git@github.com:QubesOS/build-logs') %}
+{% set build_bot_name = salt('pillar.get')('build-infra:build_bot_name', 'Qubes OS build bot') %}
+{% set build_bot_email = salt('pillar.get')('build-infra:build_bot_email', 'builder-bot@qubes-os.org') %}
```
Comment (marmarek) [inline build-infra/build-logs.sls]: This fails as below. It should be `salt['pillar.get'](...)`.
<details>

```
  build-logs:
      - Rendering SLS 'base:build-infra.build-logs' failed: Jinja error: 'AliasedLoader' object is not callable
        Traceback (most recent call last):
          File "/usr/lib/python3/dist-packages/salt/utils/templates.py", line 393, in render_jinja_tmpl
            output = template.render(**decoded_context)
          File "/usr/lib/python3/dist-packages/jinja2/asyncsupport.py", line 76, in render
            return original_render(self, *args, **kwargs)
          File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 1008, in render
            return self.environment.handle_exception(exc_info, True)
          File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 780, in handle_exception
            reraise(exc_type, exc_value, tb)
          File "/usr/lib/python3/dist-packages/jinja2/_compat.py", line 37, in reraise
            raise value.with_traceback(tb)
          File "<template>", line 1, in top-level template code
        TypeError: 'AliasedLoader' object is not callable
        
        ; line 1
        
        ---
        {% set build_log_key_fpr = salt('pillar.get')('build-infra:build_log_key_fpr', '1B760CED53D8EB5AE529BEBB16D7539A228D30DB') %}    <======================
        {% set build_log_repo = salt('pillar.get')('build-infra:build_log_repo', 'git@github.com:QubesOS/build-logs') %}
        {% set build_bot_name = salt('pillar.get')('build-infra:build_bot_name', 'Qubes OS build bot') %}
        {% set build_bot_email = salt('pillar.get')('build-infra:build_bot_email', 'builder-bot@qubes-os.org') %}
        
        # logs signing key, secret key is needed too (not configured through salt)
        [...]
        ---
```

File: build-infra/build-vm.sls [hunk 6]
```diff
@@ -63,12 +69,27 @@ gpg --import /home/user/qubes-master-key.asc:
     - onchange:
       - file: /home/user/qubes-master-key.asc
 
+{{qubes_master_key_fpr}}:
+  gpg.present:
+    - user: user
```
Comment (marmarek) [inline build-infra/build-vm.sls:74]: This fails as below. Most likely missing package. `python3-gnupg`?
<details>

```
  build-logs:
      - Rendering SLS 'base:build-infra.build-logs' failed: Jinja error: 'AliasedLoader' object is not callable
        Traceback (most recent call last):
          File "/usr/lib/python3/dist-packages/salt/utils/templates.py", line 393, in render_jinja_tmpl
            output = template.render(**decoded_context)
          File "/usr/lib/python3/dist-packages/jinja2/asyncsupport.py", line 76, in render
            return original_render(self, *args, **kwargs)
          File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 1008, in render
            return self.environment.handle_exception(exc_info, True)
          File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 780, in handle_exception
            reraise(exc_type, exc_value, tb)
          File "/usr/lib/python3/dist-packages/jinja2/_compat.py", line 37, in reraise
            raise value.with_traceback(tb)
          File "<template>", line 1, in top-level template code
        TypeError: 'AliasedLoader' object is not callable
        
        ; line 1
        
        ---
        {% set build_log_key_fpr = salt('pillar.get')('build-infra:build_log_key_fpr', '1B760CED53D8EB5AE529BEBB16D7539A228D30DB') %}    <======================
        {% set build_log_repo = salt('pillar.get')('build-infra:build_log_repo', 'git@github.com:QubesOS/build-logs') %}
        {% set build_bot_name = salt('pillar.get')('build-infra:build_bot_name', 'Qubes OS build bot') %}
        {% set build_bot_email = salt('pillar.get')('build-infra:build_bot_email', 'builder-bot@qubes-os.org') %}
        
        # logs signing key, secret key is needed too (not configured through salt)
        [...]
        ---
```

**Review by marmarek — CHANGES_REQUESTED** [review]

File: build-infra/build-vm.sls [hunk 7]
```diff
@@ -144,3 +165,109 @@ commands-keyring:
...
+
+{{builder}}-get:
+  git.latest:
+    - name : https://github.com/QubesOS/qubes-builder
+    - target: {{ builder }}
+    - user: user
+
+{{builder}}-check:
+  cmd.run:
+    - name: git verify-tag --raw "$(git describe --abbrev=0)" 2>&1 >/dev/null | grep '^\[GNUPG:\] TRUST_FULLY'
```
Comment (marmarek) [inline build-infra/build-vm.sls]: `--abbrev=0` here is wrong. It will verify latest available tag, not the tag describing that is downloaded. In other words - it won't fail if some malicious commits are added at the top.

File: pillar/build-infra/build-logs.sls [hunk 8]
```diff
@@ -1,7 +1,13 @@
...
+# public gpg key associated to the one used to sign log commits
+# it allows to not rely on gpg key servers
+  build_log_public_key: |
+    -----BEGIN PGP PUBLIC KEY BLOCK-----
+    ...
+    -----END PGP PUBLIC KEY BLOCK-----
 # ssh key used to access github - needs to have write access to build-logs repository
   github_ssh_key: |
-    -----BEGIN RSA PRIVATE KEY-----
+    -----BEGIN OPENSSH PRIVATE KEY-----
```
Comment (marmarek) [inline pillar/build-infra/build-logs.sls]: Is this change intentional? Since the key is put in `id_rsa`, I think it is expected to be RSA.

File: pillar/build-infra/init.sls [hunk 9]
```diff
@@ -1,21 +1,61 @@
...
+            - 9FA64B92F95E706BF28E2CA6484010B5CDC576E2
+          release: 4.0
+          config:
+            - file: R4.0/qubes-os-r4.0-dom0.conf
+            - repository:
+                - baseurl: https://github.com/QubesOS-contrib/qubes-
+                - component: contrib-configs
+        /home/user/builder-r4.1:
+          release: 4.1
+          config:
```
Comment (marmarek) [inline pillar/build-infra/init.sls]: Trailing space, here and in few lines below.

**Review by marmarek — CHANGES_REQUESTED** [review]

File: pillar/build-infra/build-logs.sls (hunk 8 above)
Comment (fepitre) [reply pillar/build-infra/build-logs.sls]: I thought it was default to have OPENSSH. But well, it seems not.

File: build-infra/build-vm.sls (hunk 7 above)
Comment (fepitre) [reply build-infra/build-vm.sls]: Should it be: ```git verify-commit --raw HEAD```?

Comment (marmarek) [reply build-infra/build-vm.sls]: verify-commit and verify-tag are two different things. It should be tag here (otherwise contributor's commit on top, only signed with my tag wouldn't be accepted).

**Review by marmarek — APPROVED** [review]
